syntax = "proto3";

message ResponseEnvelop {
  int32 unknown1 = 1; // Always 1, besides first response which is 53.
  int64 rpc_id = 2; // Random
  string endpoint = 3; // Only in first response. The endpoint to sent future requests to.
  Unknown6 unknown6 = 6;
  TokenData token = 7; // A token received on every response. Doesn't change in session.
  repeated bytes responses = 100;

  message Unknown6 {
    int32 unknown1 = 1;
    Unknown2 unknown2 = 2;

    message Unknown2 {
      int32 unknown1 = 1;
    }
  }

  message TokenData {
    bytes token = 1;
    int64 timestamp = 2;
    bytes sig = 3; // Possibly a SHA1 signature
  }
}

message PossibleResponse1 { // GET_MAP_OBJECTS response?
  repeated MapTile map_tiles = 1;
  int32 Status = 2;

  message MapTile {
    int64 map_cell_id = 1;
    int64 timestamp = 2;
    repeated Fort forts = 3;
    repeated ClientSpawnPoint spawn_point = 4;
    repeated WildPokemon wild_pokemon = 5;
    int32 deleted_object = 6;
    int32 is_truncated_list = 7;
    int32 fort_summary = 8;
    ClientSpawnPoint decimated_spawn_point = 9;
    repeated MapPokemon catchable_pokemon = 10;
    repeated NearbyPokemon nearby_pokemon = 11;

    message NearbyPokemon {
      int32 pokedex_number = 1;
      fixed32 distance_meters = 2;
      fixed64 encounter_id = 3;
    }

    // Pokestop or Gym
    message Fort {
      string fort_id = 1;
      int64 last_modified_time = 2;
      fixed64 lat = 3;
      fixed64 lng = 4;
      int32 team = 5; // blue=1, red=2, yellow=3
      int32 guard_pokemon_id = 6;
      int32 guard_pokemon_level = 7;
      int32 enabled = 8;
      int32 fort_type = 9; // If this appears, it's a pokestop
      int32 gym_points = 10; // prestige
      int32 active_module = 11;
      bytes unknown12 = 12;
      LureInfo lure = 13;
      int64 cooldown_complete_time = 14;

      message LureInfo {
        bytes pokestop_id = 1;
        fixed64 unknown2 = 2;
        int32 unknown3 = 3; // apprently the pokemon id that will appear
        int64 expire_time = 4;
      }
    }

    message ClientSpawnPoint {
      fixed64 lat = 2;
      fixed64 lng = 3;
    }

    message WildPokemon {
      fixed64 encounter_id = 1;
      int64 last_modified_time = 2;
      fixed64 lat = 3;
      fixed64 lng = 4;
      string spawn_point_id = 5;
      PokeInfo info = 7;
      int64 time_till_hidden = 11;

      message PokeInfo {
        int32 number = 2;
      }
    }

    message MapPokemon {
      string spawn_point_id = 1;
      fixed64 encounter_id = 2;
      int32 pkmn_no = 3;
      int32 time = 4; //disapper time
      fixed64 lat = 5;
      fixed64 lng = 6;
    }
  }
}

message PossibleResponse2 { // ??
  int32 unknown1 = 1;
  Unknown2 unknown2 = 2;

  message Unknown2 {
    int32 unknown1 = 1; // Looks like timestamp
    int32 unknown2 = 2; // Looks like timestamp. Always the same as 1.
  }
}

message PossibleResponse3 { // ??
  string unknown2 = 2;
}

message PossibleResponse4 { // ?? pm0XXX response
  repeated Unknown1 unknown1 = 1;
  int64 unknown2 = 2; // Timestamp in milliseconds.

  message Unknown1 {
    string id = 1;
    bytes pms = 2;
    int64 timestamp = 3; // Milliseconds.
    int32 unknown4 = 4;
    int32 unknown5 = 5;
    string unknown6 = 6;
  }
}

//message PossibleResponse5 { // GET_PLAYER response?
//  int32 unknown1 = 1;
//  PlayerDetails player = 2;
//}


